ArduCopter/AP_Arming.cpp

```diff
@@ -650,6 +650,13 @@ bool AP_Arming_Copter::arm_checks(AP_Arming::Method method)
         return false;
     }
 
+    // always check if the takeoff is the first command for a mission in auto mode
+    if (copter.get_mode() == (uint8_t)Mode::Number::AUTO) {
+      if (!copter.mode_auto.mode_auto_pre_arm_check()) {
+        check_failed(true, "First command should be Takeoff");
```

Comment:
This is making too much of an assumption about what's going on in the mode to disallow arming.

This should simply be 
`if (!copter.flightmode->pre_arm_check(report))`

This should also be a pre-arm check rather than an arming check.

The flightmode base class should have `pre_arm_check(bool report)` on it as a virtual class, with auto overriding it.

If it fails it should call `AP::arming().check_failed(...)`.

Reply:
got it!I will start working on it sir.Thanks